File: src/main/java/org/xodium/vanillaplus/commands/ReloadCommand.java
package org.xodium.vanillaplus.commands;

import java.util.List;

import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.xodium.vanillaplus.VanillaPlus;
import org.xodium.vanillaplus.gui.SettingsGUI;

import com.mojang.brigadier.Command;
import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;

/**
 * The ReloadCommand class handles the registration and execution of the
 * "vanillaplus" command, which opens the settings GUI for players with the
 * appropriate permissions.
 * 
 * <p>
 * This command can only be executed by players. If a player does not have
 * the required permission, they will receive an error message.
 * </p>
 * 
 * <p>
 * Constants:
 * </p>
 * <ul>
 * <li>{@code VP} - An instance of the VanillaPlus plugin.</li>
 * <li>{@code MM} - An instance of the MiniMessage parser.</li>
 * </ul>
 * 
 * <p>
 * Interfaces:
 * </p>
 * <ul>
 * <li>{@code MSG} - Contains pre-defined messages for permission errors and
 * player-only command errors.</li>
 * <li>{@code PERMS} - Contains the permission string required to execute the
 * reload command.</li>
 * </ul>
 * 
 * <p>
 * Static Initialization Block:
 * </p>
 * <ul>
 * <li>Registers the "vanillaplus" command with the command registrar during
 * the plugin's lifecycle events.</li>
 * <li>Checks if the command sender is a player and has the required
 * permission before opening the settings GUI.</li>
 * <li>Sends appropriate error messages if the sender is not a player or
 * lacks the required permission.</li>
 * </ul>
 */
public class ReloadCommand {
    private static final VanillaPlus VP = VanillaPlus.getInstance();
    private static final MiniMessage MM = MiniMessage.miniMessage();

    private static interface MSG {
        @NotNull
        Component PERM_ERR = MM.deserialize(VanillaPlus.PREFIX
                + "<red>You do not have permission to use this command!");
        @NotNull
        Component PLAYER_ONLY_CMD = MM
                .deserialize("This command can only be run by a player.");
    }

    private static interface PERMS {
        String RELOAD = VP.getClass().getSimpleName() + ".reload";
    }

    static {
        VP.getLifecycleManager().registerEventHandler(LifecycleEvents.COMMANDS, e -> {
            e.registrar().register(
                    Commands.literal("vanillaplus")
                            .executes(ctx -> {
                                CommandSender cs = ctx.getSource().getSender();
                                if (cs instanceof Player p) {
                                    if (!p.hasPermission(PERMS.RELOAD)) {
                                        p.sendMessage(MSG.PERM_ERR);
                                        return 0;
                                    }
                                    SettingsGUI.openInventory(p);
                                    return Command.SINGLE_SUCCESS;
                                }
                                cs.sendMessage(MSG.PLAYER_ONLY_CMD);
                                return 0;
                            })
                            .build(),
                    "Opens the GUI",
                    List.of("vp"));
        });
    }
}

File: src/main/kotlin/org/xodium/vanillaplus/commands/ReloadCommand.kt
/*
 *  Copyright (c) 2025. Xodium.
 *  All rights reserved.
 */

@file:Suppress("UnstableApiUsage")

package org.xodium.vanillaplus.commands

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.Commands
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents
import org.xodium.vanillaplus.Utils
import org.xodium.vanillaplus.VanillaPlus
import org.xodium.vanillaplus.VanillaPlus.Companion.instance


/**
 * The `ReloadCommand` object represents a command handler for reloading the plugin.
 * This command is registered during the plugin's lifecycle and can be executed by players
 * or the console to reload the plugin without restarting the server.
 *
 * Upon successful execution, it notifies the sender with a confirmation message and logs
 * the reload action to the server console.
 */
object ReloadCommand {
    init {
        instance.lifecycleManager.registerEventHandler(LifecycleEvents.COMMANDS) {
            it.registrar().register(
                Commands.literal("vanillaplus")
                    .executes(Command { ctx: CommandContext<CommandSourceStack?>? ->
                        val cs = ctx?.source?.sender ?: return@Command 0
                        try {
//                            TODO: Implement reload logic
//                            val pluginManager = instance.server.pluginManager
//                            pluginManager.disablePlugin(instance)
//                            pluginManager.enablePlugin(instance)
                            instance.logger.info("Reloaded successfully")
                            cs.sendMessage(Utils.MM.deserialize("${VanillaPlus.PREFIX}<green>Reloaded successfully."))
                        } catch (e: Exception) {
                            instance.logger.severe("Failed to reload: ${e.message}")
                            cs.sendMessage(Utils.MM.deserialize("${VanillaPlus.PREFIX}<red>Failed to reload. Check server logs for details."))
                        }
                        Command.SINGLE_SUCCESS
                    })
                    .build(),
                "Reloads the VanillaPlus plugin",
                mutableListOf("vp")
            )
        }
    }
}
